### PR TITLE
First version of the linker

### PR DESCRIPTION
**Pull request by vlopezferrando**
Implemented the first version of the linker does the following:

1. Read a single CSV file which contains a column named `bf` with a string of 1s and 0s.
2. Generate reference bloom filters by one of two methods: A) random values, B) picking one of the CSV bloom filters.
3. Find $k^*$, the $k$ for which the $k$-means clustering algorithms produces labels that reduce *purity*.
4. Use $k^*$ to run $k$-means clustering, label the original dataset and save it as another CSV file.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,16 @@
 # linkage
 Implementation of the Linkage algorithm that pairs with the Tokenizer
+
+## Installation
+
+Create a virtual environment:
+
+    python3.11 -m venv venv
```
Comment (sanathkr) [inline README.md]: Are we pinned to 3.11 version somehow? If not, we should rename this to use just `python`.

File: linker.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import itertools
+
+import numpy as np
+import pandas as pd
+from sklearn.cluster import KMeans
+
+
+RANDOM_SEED = 0
```
Comment (sanathkr) [inline linker.py:11]: ah man, we gotta get geeky here. How about `46` for the 46 chromosomes we all have and the cause of all genetic diseases? :)

File: linker.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            max_purity, k_star = purity_k, k
+
+    print(f"\nk*={k_star}, purity={max_purity:.2f}")
+    return k_star
+
+
+def run():
+    np.random.seed(RANDOM_SEED)
+
+    fname = "data/sample.csv"
```
Comment (sanathkr) [inline linker.py]: eventually we should make this dynamic.

Also, the `data` folder doesn't seem to be checked into the PR. Should you also publish that?

File: linker.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    # Check that CSV file has column `bf`
+    assert "bf" in df.columns, "CSV file doesn't have column `bf`"
+
+    # Get bloom filter length and check that all bloom filters have the same length
+    l = len(df.bf[0])
+    assert all(df.bf.str.len() == l), "CSV file has different length of bloom filters"
+
+    # Convert bloom filters to numpy arrays
+    D = df.bf.apply(lambda x: np.array(list(map(int, list(x)))))
```
Comment (sanathkr) [inline linker.py:112]: we can get more descriptive with variable names? I don't want this to become a research code that ends up becoming really hard to read. I loved your Typescript names. Wouldn't mind a bit more verbosity here :)

**Review by sanathkr — CHANGES_REQUESTED** [review]
I don't know the algorithm very well, so we should review this line by line over a call. I am also going to push for better variable names. I know this implementation mimics equations in the paper. For next iteration, it will be worth expanding the variables and making the implementation a bit more readable.

File: README.md (hunk 1 above)
Comment (vlopezferrando) [reply README.md]: Not really, I have changed it to `python3`.

File: linker.py (hunk 2 above)
Comment (vlopezferrando) [reply linker.py:11]: Haha, there is a reason why I chose 0: I was testing a sample and the correct result was 4. With 0, the result was wrong, it gave 6, but basically with any other seed I got 4.

I thought changing the 0 would be considered cheating, then :D.

**Review by sanathkr — APPROVED** [review]
